File: dcr/embed.py
"""L2 vectors without a model dependency.

A hashing embedder (the "hashing trick") is enough for the runtime's actual
requirement: making state nodes *findable by meaning-ish similarity* so the
planner has seeds to expand from. It is deterministic, dependency-free and
instant, which keeps the whole runtime runnable offline. Swap in real
embeddings by passing any callable of `str -> list[float]` to the Ladder.
"""

from __future__ import annotations

import math
import re
from collections import Counter

TOKEN_RE = re.compile(r"[A-Za-z0-9_.:/-]+")
DIM = 256

STOPWORDS = frozenset(
    """a an and are as at be by for from has have in is it its of on or that the
    to was were will with what which who whom this these those do does did not
    you your we our they their he she his her i me my but if then than so""".split()
)

INSTRUCTION_WORDS = frozenset(
    """quote exactly exact verbatim literal tell show give me please line text
    word words message wording""".split()
)
"""Words that describe the *form* of an answer rather than its content.
Scoring on them drags every query toward whichever line happens to contain
"message" or "exactly", so they are stripped alongside stopwords."""

SKIP = STOPWORDS | INSTRUCTION_WORDS


def tokenize(text: str) -> list[str]:
    return [t.lower() for t in TOKEN_RE.findall(text or "")]


def content_tokens(text: str) -> list[str]:
    return [t for t in tokenize(text) if t not in SKIP and len(t) > 1]


def hashing_embed(text: str, dim: int = DIM) -> list[float]:
    """Sub-word + word hashing embedding, L2-normalised."""
    vec = [0.0] * dim
    tokens = content_tokens(text)
    if not tokens:
        return vec
    counts = Counter(tokens)
    for token, count in counts.items():
        weight = 1.0 + math.log(count)
        vec[hash_index(token, dim)] += weight
        # Character trigrams give partial credit for morphological variants
        # and typos, which plain word hashing cannot do.
        for i in range(len(token) - 2):
            vec[hash_index(token[i : i + 3], dim)] += 0.35 * weight
    norm = math.sqrt(sum(v * v for v in vec))
    if norm:
        vec = [v / norm for v in vec]
    return vec


def hash_index(token: str, dim: int = DIM) -> int:
    # Python's builtin hash() is salted per process; use a stable digest so
    # persisted vectors stay comparable across runs.
    h = 2166136261
    for ch in token.encode("utf-8"):
        h = ((h ^ ch) * 16777619) & 0xFFFFFFFF
    return h % dim


def cosine(a: list[float], b: list[float]) -> float:
    if not a or not b:
        return 0.0
    return sum(x * y for x, y in zip(a, b))

File: src/embed.rs
//! L2 vectors without a model dependency.
//!
//! A hashing embedder is enough for the runtime's actual requirement: making
//! state nodes findable by meaning-ish similarity so the planner has seeds to
//! expand from. Deterministic, dependency-free and instant, which keeps the
//! whole runtime runnable offline.

use std::collections::HashMap;

use crate::text::content_tokens;

pub const DIM: usize = 256;

/// FNV-1a over the token, so persisted vectors stay comparable across runs.
fn hash_index(token: &str, dim: usize) -> usize {
    let mut hash: u32 = 2_166_136_261;
    for byte in token.as_bytes() {
        hash ^= *byte as u32;
        hash = hash.wrapping_mul(16_777_619);
    }
    (hash as usize) % dim
}

/// Word + character-trigram hashing embedding, L2-normalised.
///
/// The trigrams give partial credit for morphological variants and typos,
/// which plain word hashing cannot do.
pub fn hashing_embed(text: &str, dim: usize) -> Vec<f32> {
    let mut vec = vec![0.0f32; dim];
    let tokens = content_tokens(text);
    if tokens.is_empty() {
        return vec;
    }
    let mut counts: HashMap<&str, u32> = HashMap::new();
    for token in &tokens {
        *counts.entry(token.as_str()).or_insert(0) += 1;
    }
    for (token, count) in counts {
        let weight = 1.0 + (count as f32).ln();
        vec[hash_index(token, dim)] += weight;
        let chars: Vec<char> = token.chars().collect();
        if chars.len() >= 3 {
            for window in chars.windows(3) {
                let trigram: String = window.iter().collect();
                vec[hash_index(&trigram, dim)] += 0.35 * weight;
            }
        }
    }
    let norm = vec.iter().map(|v| v * v).sum::<f32>().sqrt();
    if norm > 0.0 {
        for v in &mut vec {
            *v /= norm;
        }
    }
    vec
}

pub fn embed(text: &str) -> Vec<f32> {
    hashing_embed(text, DIM)
}

pub fn cosine(a: &[f32], b: &[f32]) -> f32 {
    if a.is_empty() || b.is_empty() {
        return 0.0;
    }
    a.iter().zip(b).map(|(x, y)| x * y).sum()
}

/// Anything that turns text into a vector.
///
/// The documentation has claimed since the first release that "the interface
/// accepts any `str -> Vec<f32>`". It did not: `Ladder` called
/// [`hashing_embed`] directly, so swapping the embedder meant editing the
/// runtime. This is that claim made true.
///
/// The default stays [`HashingEmbedder`], because the whole runtime is offline,
/// deterministic and dependency-free, and a learned embedder is none of those.
/// What changes is that replacing it is now a constructor argument rather than
/// a patch.
pub trait Embedder: std::fmt::Debug {
    fn embed(&self, text: &str) -> Vec<f32>;

    /// Dimensionality. Vectors from one embedder are never comparable with
    /// another's, so this is checked rather than assumed where it matters.
    fn dim(&self) -> usize;
}

/// The bundled embedder: word + character-trigram hashing, 256 dimensions.
///
/// Fast, offline and reproducible across runs. It finds material that shares
/// vocabulary, which is why paraphrase retrieval is a documented poor fit —
/// `bench --channels` measures how much this channel actually agrees with the
/// lexical one rather than leaving it asserted.
#[derive(Debug, Clone, Copy)]
pub struct HashingEmbedder {
    pub dim: usize,
}

impl Default for HashingEmbedder {
    fn default() -> Self {
        Self { dim: DIM }
    }
}

impl Embedder for HashingEmbedder {
    fn embed(&self, text: &str) -> Vec<f32> {
        hashing_embed(text, self.dim)
    }

    fn dim(&self) -> usize {
        self.dim
    }
}
